src/altrep-sparse-real.c

```diff
@@ -0,0 +1,176 @@
+#define R_NO_REMAP
+#include "R.h"
+#include <Rinternals.h>
+#include <R_ext/Altrep.h>
+
+// Initialised at load time
+R_altrep_class_t altrep_sparse_real_class;
+
+SEXP ffi_altrep_new_sparse_real(SEXP x) {
+  return R_new_altrep(altrep_sparse_real_class, x, R_NilValue);
+}
+
+SEXP alrep_sparse_real_Materialize(SEXP vec) {
+  SEXP out = R_altrep_data2(vec);
+
+  if (out != R_NilValue) {
+    return out;
+  }
+
+  SEXP data1 = R_altrep_data1(vec);
+  SEXP val = VECTOR_ELT(data1, 0);
+  SEXP pos = VECTOR_ELT(data1, 1);
+  SEXP len = VECTOR_ELT(data1, 2);
+
+  out = PROTECT(Rf_allocVector(REALSXP, Rf_asInteger(len)));
+  
+  int n = Rf_asInteger(len);
+  
+  for (int i = 0; i < n; ++i) {
+    SET_REAL_ELT(out, i, 0);
+  }
+
+  int m = Rf_length(pos);
+
+  for (int i = 0; i < m; ++i) {
+    SET_REAL_ELT(out, INTEGER_ELT(pos, i) - 1, REAL_ELT(val, i));
+  }
+
+  R_set_altrep_data2(vec, out);
+
+  UNPROTECT(1);
+  return out;
+}
+
+// -----------------------------------------------------------------------------
+// ALTVEC
+
+void* altrep_sparse_real_Dataptr(SEXP x, Rboolean writeable) {
+  return STDVEC_DATAPTR(alrep_sparse_real_Materialize(x));
+}
+
+const void* altrep_sparse_real_Dataptr_or_null(SEXP vec) {
+    SEXP out = R_altrep_data2(vec);
+
+  if (out == R_NilValue) {
+    return NULL;
+  } else {
+    return out;
+  }
+}
+
+static SEXP altrep_sparse_real_Extract_subset(SEXP x, SEXP indx, SEXP call) {
+
+  SEXP data1 = R_altrep_data1(x);
+  SEXP val_old = VECTOR_ELT(data1, 0);
+  SEXP pos_old = VECTOR_ELT(data1, 1);
+  SEXP matches = Rf_match(pos_old, indx, R_NaInt);
```

Comment:
```suggestion
  SEXP matches = PROTECT(Rf_match(pos_old, indx, R_NaInt));
```

This is currently unsafe, `Rf_match()` allocates a new vector of matches, so you must protect it.

Also requires bumping the `UNPROTECT()` count up by 1